service/worker/failovermanager/workflow.go

```diff
@@ -0,0 +1,380 @@
+		SuccessDomains []string
+		FailedDomains  []string
+	}
+
+	// QueryResult for failover progress
+	QueryResult struct {
+		TotalDomains   int
+		Success        int
+		Failed         int
+		State          string
+		TargetCluster  string
+		SourceCluster  string
+		SuccessDomains []string
+		FailedDomains  []string
+	}
+)
+
+func init() {
+	workflow.RegisterWithOptions(FailoverWorkflow, workflow.RegisterOptions{Name: WorkflowTypeName})
+	activity.RegisterWithOptions(FailoverActivity, activity.RegisterOptions{Name: failoverActivityName})
+	activity.RegisterWithOptions(GetDomainsActivity, activity.RegisterOptions{Name: getDomainsActivityName})
+}
+
+// FailoverWorkflow is the workflow that managed failover all domains with IsManagedByCadence=true
+func FailoverWorkflow(ctx workflow.Context, params *FailoverParams) (*FailoverResult, error) {
+	err := validateParams(params)
+	if err != nil {
+		return nil, err
+	}
+
+	// define query properties
+	var failedDomains []string
+	var successDomains []string
+	var totalNumOfDomains int
+	wfState := WorkflowRunning
+	err = workflow.SetQueryHandler(ctx, QueryType, func(input []byte) (*QueryResult, error) {
+		return &QueryResult{
+			TotalDomains:   totalNumOfDomains,
+			Success:        len(successDomains),
+			Failed:         len(failedDomains),
+			State:          wfState,
+			TargetCluster:  params.TargetCluster,
+			SourceCluster:  params.SourceCluster,
+			SuccessDomains: successDomains,
+			FailedDomains:  failedDomains,
+		}, nil
+	})
+	if err != nil {
+		return nil, err
+	}
+
+	// get target domains
+	ao := workflow.WithActivityOptions(ctx, getGetDomainsActivityOptions())
+	getDomainsParams := &GetDomainsActivityParams{
+		TargetCluster: params.TargetCluster,
+		SourceCluster: params.SourceCluster,
+		Domains:       params.Domains,
+	}
+	var domains []string
+	err = workflow.ExecuteActivity(ao, GetDomainsActivity, getDomainsParams).Get(ctx, &domains)
+	if err != nil {
+		return nil, err
+	}
+	totalNumOfDomains = len(domains)
+
+	// failover in batch
+	ao = workflow.WithActivityOptions(ctx, getFailoverActivityOptions())
+	batchSize := params.BatchFailoverSize
+	times := len(domains)/batchSize + 1
+
+	pauseCh := workflow.GetSignalChannel(ctx, PauseSignal)
+	resumeCh := workflow.GetSignalChannel(ctx, ResumeSignal)
+	var shouldPause bool
+
+	for i := 0; i < times; i++ {
+		// check if need to pause
+		shouldPause = pauseCh.ReceiveAsync(nil)
+		if shouldPause {
+			wfState = WorkflowPaused
+			resumeCh.Receive(ctx, nil)
+		}
+		wfState = WorkflowRunning
+
+		failoverActivityParams := &FailoverActivityParams{
+			Domains:       domains[i*batchSize : common.MinInt((i+1)*batchSize, totalNumOfDomains)],
+			TargetCluster: params.TargetCluster,
+		}
+		var actResult FailoverActivityResult
+		err = workflow.ExecuteActivity(ao, FailoverActivity, failoverActivityParams).Get(ctx, &actResult)
+		if err != nil {
+			failedDomains = append(failedDomains, failoverActivityParams.Domains...)
```

Comment:
I think even if the activity has failed, part of the domains may still be successfully failovered to the target cluster. Then if we do a rollback, not all domains will be rollbacked. We probably need a third unknown state for the domains. And also rollback those unknown domains, which will either be a no-op or be failovered to the source cluster.

Reply:
Good point, you are right the current rollback may falsely left domains during rollback.  
I want to keep it simple, so I didn't update here with more state, but let rollback handle all domains (both success and failed), the new start GetDomainActivity will do correct things. 
I also updated comments on query result FailedDomains to let caller know it may contains False positive. I suppose operator will always redo failover for FailedDomains so it may acceptable to contain false positive results.